Add complex dtype and isreal function - Ivy functional API extension

Nirzu97 opened the pull request:
closes #4789

Nirzu97 commented:
I understand that adding complex dtype will fail many tests where helpers are drawing data from "valid dtypes" or "numeric dtypes". Should I make changes in helpers? Or changes in how I add complex type?

For example, I can remove complex data types from numeric and valid and make new helper just for complex data type. so it can be used only when complex numbers are needed and wouldn't affect other tests for now.

Nirzu97 commented:
> I understand that adding complex dtype will fail many tests where helpers are drawing data from "valid dtypes" or "numeric dtypes". Should I make changes in helpers? Or changes in how I add complex type?
> 
> For example, I can remove complex data types from numeric and valid and make new helper just for complex data type. so it can be used only when complex numbers are needed and wouldn't affect other tests for now.

@VedPatwardhan and @CatB1t I saw your many commits on the support of different datatypes. Can you please take a look at this and suggest what should I do in this case

VedPatwardhan commented on ivy/__init__.py:
```diff
@@ -224,6 +249,9 @@ class Node(str):
     float16,
     float32,
     float64,
+    complex64,
```
I think we shouldn't have complex dtypes in the numeric category as most of the tests which are generating data with numeric datatypes could fail as they may not support complex dtypes. The same applies to all other files changed.

VedPatwardhan commented on ivy/functional/backends/numpy/__init__.py:
```diff
@@ -36,6 +36,8 @@
 float32 = ivy.FloatDtype("float32")
 float64 = ivy.FloatDtype("float64")
+complex64 = ivy.ComplexDtype("complex64")
+complex128 = ivy.ComplexDtype("complex128")
```
NumPy also supports `complex256`, so please add it to the `numpy` backend.

Nirzu97 replied:
@VedPatwardhan I added it initially but it seems it is not supported in v1.23.0:

>>>
>>> np.__version__
'1.23.0'
>>> np.complex64   
<class 'numpy.complex64'>
>>> np.complex128 
<class 'numpy.complex128'>
>>> np.complex256
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\HP\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\__init__.py", line 311, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'complex256'

Nirzu97 replied:
@VedPatwardhan I added it initially and it is supported on Linux. But Numpy doesn't support `complex256` on Windows. it throws the following error for all the files in numpy backend: `'AttributeError: module 'numpy' has no attribute 'complex256'` 

Should I still add it in `numpy` backend?

VedPatwardhan commented on ivy_tests/array_api_methods_to_test/elementwise_functions.txt:
```diff
@@ -54,3 +54,4 @@ subtract
 tan
 tanh
 trunc
+isreal
```
`isreal` isn't a part of the array api standard, so it shouldn't be added here

VedPatwardhan reviewed (changes requested):
Hi @Nirzu97, thanks for contributing! I've suggested a few changes to your PR to fix a few changes, apart from that it seems good to me. Feel free to request a review from @JLU-Neal once you're done with the changes, I think provided there aren't any large scale failures, we could merge it. Thanks 🙂

Nirzu97 commented:
@VedPatwardhan I have fixed the failing tests and made other changes. I requested a review but it doesn't have the option to request a review from @JLU-Neal from my side. Also, he seems extremely busy with other work. Can you please check this or assign it to someone else?

VedPatwardhan reviewed (changes requested):
Hi @Nirzu97, thanks for making the changes, there are quite a few changes to be made because the structure of the helpers has changed, etc. I have made those final changes on my local system but I'm not able to push it to your branch, probably because of some issues with permissions, etc. so I've created a temporary pull request [here](https://github.com/unifyai/ivy/pull/4974/files) which you could check out to replicate it on your PR. Following are the main points regarding the same:
1. The array api standard doesn't support complex dtypes yet, so I've removed them from the array api promotion table and added to the extra promotion table instead.
2. Let's avoid type promotions between numeric and complex types for now as it isn't related to the extension you're adding.
3. I've removed the complex types from the valid dtypes because valid dtypes are being used for testing a number of functions and adding complex dtypes there could cause failures in other functions.
4. I've updated the `get_dtypes` helper in the `dtype_helpers.py` file to handle complex dtypes properly.
5. I've also updated the `array_values` helper to function properly for your function.
6. I've updated the helper usage in the `test_isreal` function according to the updated format of returns from those helpers.
7. Finally, I've removed changes made to files which are unrelated to your task.

Feel free to re-request my review once you're done with those changes. Thanks 🙂

Nirzu97 commented:
> Hi @Nirzu97, thanks for making the changes, there are quite a few changes to be made because the structure of the helpers has changed, etc. I have made those final changes on my local system but I'm not able to push it to your branch, probably because of some issues with permissions, etc. so I've created a temporary pull request [here](https://github.com/unifyai/ivy/pull/4974/files) which you could check out to replicate it on your PR. Following are the main points regarding the same:
> 
> 1. The array api standard doesn't support complex dtypes yet, so I've removed them from the array api promotion table and added to the extra promotion table instead.
> 2. Let's avoid type promotions between numeric and complex types for now as it isn't related to the extension you're adding.
> 3. I've removed the complex types from the valid dtypes because valid dtypes are being used for testing a number of functions and adding complex dtypes there could cause failures in other functions.
> 4. I've updated the `get_dtypes` helper in the `dtype_helpers.py` file to handle complex dtypes properly.
> 5. I've also updated the `array_values` helper to function properly for your function.
> 6. I've updated the helper usage in the `test_isreal` function according to the updated format of returns from those helpers.
> 7. Finally, I've removed changes made to files which are unrelated to your task.
> 
> Feel free to re-request my review once you're done with those changes. Thanks 🙂

I have made the required changes

Nirzu97 commented:
@VedPatwardhan I have mirrored the changes from temporary PR. Please review this.

VedPatwardhan commented on ivy_tests/array_api_methods_to_test/elementwise_functions.txt:
```diff
@@ -53,4 +53,4 @@ sqrt
 tan
 tanh
-trunc
+trunc
```
Please update this file so that it doesn't appear in the files changed, as we shouldn't be making changes to the array api test suite at all

Nirzu97 replied:
@VedPatwardhan I have not changed anything in it. It is same as [master](https://github.com/unifyai/ivy/blob/master/ivy_tests/array_api_methods_to_test/elementwise_functions.txt). Also when I try to remove it, [it shows same thing for the previous line](https://github.com/unifyai/ivy/pull/4811/commits/096885d09a348427d491b8f0ae1cb9527492f2ad#diff-1b93f37da8c5d9bd34902cecc90d492dab74f69db1afc6b04c68da7517443b3f).

VedPatwardhan commented on ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:
```diff
@@ -2447,3 +2447,37 @@ def test_trunc_divide(
+        instance_method=instance_method,
+        fw=fw,
+        fn_name="isreal",
+        x=np.asarray(x, dtype=input_dtype),
```
According to the update made recently, `x` received from the helper function is already a numpy array, so we don't need to use `np.asarray` here, please pass `x` directly as the keyword argument

VedPatwardhan reviewed (changes requested):
Hi @Nirzu97, thanks for making the changes, it all looks great! Just suggested a final couple of changes in the files changed and one more change would have to made inside the `function_testing.py` file inside the `_assert_dtypes_are_valid` function, where you should replace `ivy.valid_dtypes` by `ivy.valid_dtypes + ivy.valid_complex_dtypes` in the `if` condition check. Finally, please get your branch up to date with master and we could get your PR merged on the next review. 
Please feel free to ask any questions. Thanks a lot 🙂

Nirzu97 commented:
> Hi @Nirzu97, thanks for making the changes, it all looks great! Just suggested a final couple of changes in the files changed and one more change would have to made inside the `function_testing.py` file inside the `_assert_dtypes_are_valid` function, where you should replace `ivy.valid_dtypes` by `ivy.valid_dtypes + ivy.valid_complex_dtypes` in the `if` condition check. Finally, please get your branch up to date with master and we could get your PR merged on the next review. Please feel free to ask any questions. Thanks a lot 🙂

@VedPatwardhan I have made the changes and got my branch up to date with master

VedPatwardhan reviewed (approved):
lgtm! Thanks for contributing @Nirzu97 🙂
